lambdaclass/mirra_backend
# [GH-359] [CONFIG] Implement campaign config

## Motivation

Add import for main campaign levels config, as well as super_campaigns.
Closes https://github.com/lambdaclass/afk_gacha_game/issues/359

## Summary of changes

- Reuse code from Dungeon Campaign Config Import
  - Move to helper function
- Add SuperCampaign import as well
  - Reason: Necessary to allow for a way to add campaigns during runtime without needing to mess with the DB through Ecto (easier for non-technical contributors) or re-running the seeds (requires DB wipe).
- Adds new is_boss_stage boolean field to Level
- Adds new constraints to Campaign and SuperCampaign to avoid duplicates

## How to test it?

Run `Champions.Config.import_dungeon_levels_config()` and  `Champions.Config.import_main_campaign_levels_config()` (seeds run them automatically) and check that the data is inserted/updated succesfully.

Here are some queries for your convenience:

```
SELECT
	level_number,
	sc.name,
	l.experience_reward,
	l.max_units,
	l.attempt_cost,
	ch.name, 
	u.level
FROM public.levels l
JOIN public.campaigns c on l.campaign_id = c.id
JOIN public.super_campaigns sc on c.super_campaign_id = sc.id
JOIN public.units u on u.campaign_level_id = l.id
JOIN public.characters ch on u.character_id = ch.id
where sc.name = 'Dungeon'
ORDER BY level_number ASC 
```

```
SELECT
	level_number,
	sc.name,
	l.experience_reward,
	l.max_units,
	l.attempt_cost,
	curr.name as currency_reward_name,
	cr.amount as currency_reward_amount
FROM public.levels l
JOIN public.campaigns c on l.campaign_id = c.id
JOIN public.super_campaigns sc on c.super_campaign_id = sc.id
JOIN public.currency_rewards cr on cr.level_id = l.id
JOIN public.currencies curr on cr.currency_id = curr.id
where sc.name = 'Main Campaign'
ORDER BY level_number ASC 
```


```
SELECT
	level_number,
	sc.name,
	l.experience_reward,
	l.max_units,
	l.attempt_cost,
	ch.name, 
	u.level
FROM public.levels l
JOIN public.campaigns c on l.campaign_id = c.id
JOIN public.super_campaigns sc on c.super_campaign_id = sc.id
JOIN public.units u on u.campaign_level_id = l.id
JOIN public.characters ch on u.character_id = ch.id
where sc.name = 'Main Campaign'
ORDER BY level_number ASC 
```

```
SELECT
	level_number,
	sc.name,
	l.experience_reward,
	l.max_units,
	l.attempt_cost,
	curr.name as currency_reward_name,
	cr.amount as currency_reward_amount
FROM public.levels l
JOIN public.campaigns c on l.campaign_id = c.id
JOIN public.super_campaigns sc on c.super_campaign_id = sc.id
JOIN public.currency_rewards cr on cr.level_id = l.id
JOIN public.currencies curr on cr.currency_id = curr.id
where sc.name = 'Dungeon'
ORDER BY level_number ASC 
```

Unlike #663, this one should also be tested against the client, to check the campaign works as usual
